### ISSUE TITLE
[auditd] kernel panics after installation

### ISSUE THREAD

**Issue by frederikbosch** [konstruktoid/hardening#29]
First of all, great package! I have a question around auditd. Directly after installation my VM was rebooting all the time. The reason was the audit log limit exceeded and that caused a kernel panic. I found out this was caused by our backup application that was installed before I ran the hardening script.

In order to get back control of the VM I changed the failure mode from *2* to *0* in recovery mode. Now, I want to prevent the kernel panics by adding a rule to auditd. Only I have no idea what that rule should be. I already saw in the logs that the backup program (running from /usr/sbin) was doing al kinds of operations (e.g. cp, key=tmp).

What rule should I add to prevent the Kernel panic?

**Comment by frederikbosch** [comment]
@konstruktoid Thanks for your replies. I am struggling with all the rules applied to my VM, auditd and others. I would really like to work with a hardened server. But am I running against all kinds of walls. At the moment I am not even able to execute `sudo nano /etc/audit/auditd.conf`. I am only able to change things to the system in recovery mode. No idea why. I guess I am not experienced enough to have them.

**Comment by konstruktoid** [comment]
Try removing the auditd packages from the system when you're in single user/recovery mode, too see if that's the issue.

**Comment by frederikbosch** [comment]
That does not bring back the possibility to execute a command with sudo. I am on Ubuntu 18.04. `/var/log/auth.log` is telling me the following.

```log
Aug  7 15:47:26 gitlab02 sudo: pam_unix(sudo:auth): conversation failed
Aug  7 15:47:26 gitlab02 sudo: pam_unix(sudo:auth): auth could not identify password for [genkgo]
```

The last command that I was able to execute with `sudo` was the hardening script. I have removed auditd in the meanwhile, no success.

**Comment by frederikbosch** [comment]
And the password is correct, because I am able to login and change the password. I also tried to read the password from a file `sudo -S apt-get update <~/passwd.txt`. Failed too.

**Comment by frederikbosch** [comment]
@konstruktoid It seemed that I fixed both issues, not being able to run `sudo` commands and kernel panics because of auditd. Thanks for your help.

However, I am having the `sudo` problem again. As it seems, I hit the limit of 5 password failures (`deny=5` from /etc/pam.d/common-auth). When I get exited from the shell, e.g. because of non-activity due to long running script, this also counts as a deny. At least that is what I am guessing. I find this in the logs.

`machine sudo: pam_tally2(sudo:auth): user genkgo (1000) tally 8, deny 5`.

In order to fix it. I have to run recovery mode again, remove /var/log/faillog and then reboot. I have no clue how to fix it otherwise.

**Comment by frederikbosch** [comment]
I am on Ubuntu 18.04 btw.

**Comment by konstruktoid** [comment]
Great too hear that removing `auditd` fixed it; reducing the number of audit rules, using syslog/journald to send the logs to a remote server and replacing auditd halt actions with ignore might be an option as well.

You can reset the pam_tally2 locks with `pam_tally2 --user=genkgo --reset`, but this should reset automatically when you enter the correct username and password.

**Comment by frederikbosch** [comment]
> You can reset the pam_tally2 locks with pam_tally2 --user=genkgo --reset

Well, that simply does not seem to be the case. The only thing that is working is removing `/var/log/faillog`.

What did I do.

- I created a dev Docker machine (based on [solita/ubuntu-systemd:16.04](https://github.com/solita/docker-systemd)) to start playing with all scripts before using it in production again.
- Afterwards I created a user, found out logging in was a problem due to nproc, so increased nproc to 4096 which [I need anyway](https://www.elastic.co/guide/en/elasticsearch/reference/master/max-number-of-threads.html). Logging in worked.
- Then, I started logging with the wrong username/password and got blocked after 5 trials, as expected.
- Then, I tried to reset the counter with `pam_tally2 --user genkgo --reset`. But that has no consequences. You can find this out by requesting the number of failures `pam_tally2 --user genkgo`. That says 0, while at login I get `Account locked due to 9 failed logins`.
- Finally I removed `/var/log/faillog`, logging in worked again.

**Comment by frederikbosch** [comment]
And that confirms the behaviour I got yesterday on my VM.

**Comment by konstruktoid** [comment]
You're correct, but the issue is that i use 
```
sed -i '/^$/a auth required pam_tally2.so file=/var/log/faillog deny=5 unlock_time=900' "$COMMONAUTH"
``` 
which changes the default pam_tally2 log from `/var/log/tallylog` to `/var/log/faillog` (which is the pam_tally default...)

Appending `--file /var/log/faillog` to your reset will fix this issue.

Will fix so default file location is used.

**Comment by frederikbosch** [comment]
Thanks!

**Comment by frederikbosch** [comment]
I think I am able to manage. Thanks for your swift replies and help!

**Comment by frederikbosch** [comment]
@konstruktoid I am still playing with the audit backlog issue. For instance, when I ran the command `docker container prune` the audit backlog limit exceeded again. This was when I already had the number at `524288`. I guess this is due to the watch `/var/lib/docker`.

So I am wondering what the best thing to do is. Increase the number even further? Change `-f`? Why would you even want to have it at 2 (panic when a failure occurs)?

**Comment by frederikbosch** [comment]
I have asked a question regarding this topic at the [Auditd mailing list](https://www.redhat.com/archives/linux-audit/2018-August/msg00026.html). I will let you know the outcome or maybe I can create a PR to improve some configuration if necessary.

**Comment by konstruktoid** [comment]
Thanks @frederikbosch for the post to the mailing list, interesting discussion indeed, and you and Steve Grubb (https://www.redhat.com/archives/linux-audit/2018-August/msg00029.html) are both correct. 
The rules are agressive, but they are also a catch all solution and if you don't need that kind of logging the auditd settings and rules should be changed accordingly.

**Comment by frederikbosch** [comment]
@konstruktoid Clear. But I would suggest to decrease some of the aggressiveness, because my system became almost immediately unusable. As soon as systemd started my backup app, it hit the backlog limit and consequently restarted. With another machine I also used this hardening package and while I was installing an updated kernel I hit the same limit during the update. I guess we should prevent that, even if the rules are aggressive by default.

Maybe we can split the rules: `hardening.rules`, `hardening-aggressive.rules` and `hardening-docker.rules`? I guess auditd initializes these rules in order of filename, so that would make `hardening.rules` first and the others would be supplementary. What do you think about that?

**Comment by frederikbosch** [comment]
The command `aureport --start today --key --summary` gives me the following.

```
Key Summary Report
===========================
total  key
===========================
63164  tmp
16060  docker
7206  delete
6007  admin_user_home
2760  auditlog
1595  specialfiles
675  perm_mod
69  systemd
54  systemd_tools
36  init
15  sshd
12  cron
5  login
5  actions
4  access
3  privileged
1  audit_rules_networkconfig_modification
```

My suggestion would to move docker rules to a specific `hardening-docker.rules` and at least the following the keys to `hardening-aggressive.rules`: tmp, delete and admin_user_home. Maybe we can tighten the delete key a little bit more. I will create a PR for this.

**Comment by konstruktoid** [comment]
Ref the PR #30

**Comment by frederikbosch** [comment]
@konstruktoid Based on the suggestion of the audit mailing list I also updated `/tmp` and `/var/tmp` to be mounted with `noexec`. Then the rules for key `tmp` are not necessary anymore.

It only `gave` trouble with docker/compose of which I found [a solution](https://github.com/docker/compose/issues/1339#issuecomment-415068438).

Is that something you want in this package too, `noexec` for tmp?

**Comment by frederikbosch** [comment]
Sorry, I see you already suggest it in the systemd document. Maybe we should delete watching the tmp folders in the audit rules, even in the aggressive rules. The number of logs are beyond comprehension in my opinion. I doubt if that is a good thing.

**Comment by konstruktoid** [comment]
I would love to mount `/tmp` and `/var/tmp` with `noexec`, but that would break any upgrades and package installations without also modifying the behavior of `dpkg` and `apt`.

**Comment by konstruktoid** [comment]
And (after reading the new messages in the RedHat thread) the Docker rules are based on the CIS Docker Community Edition Benchmark v1.1.0.

**Comment by frederikbosch** [comment]
@konstruktoid I investigated this too. At least in Ubuntu 18.04 there is no problem anymore with `noexec` on tmp filesystems. I have done some updates/upgrades with apt, including new kernel installation, that works fine.

I read that on previous versions, 14.04 for instance, there were problems with apt, dpkg and specifically `initramfs`, and that you had to [remount](https://askubuntu.com/questions/574259/will-mounting-tmp-with-noexec-and-nosuid-cause-problems) the tmpfs in order to install updates/upgrades. I cannot tell you unfortunately which version fixed that. I have no idea.

**Comment by frederikbosch** [comment]
> And (after reading the new messages in the RedHat thread) the Docker rules are based on the CIS Docker Community Edition Benchmark v1.1.0.

Yes, I found that out too. I read version 1.13.0 and leaves me with some questions on auditing `/var/lib/docker`. It leads to so many log entries. I wonder how that can be useful. Every single change a container makes will be logged because of watching `/var/lib/docker/containers`. Therefore I applied yesterday to the CIS Workbench so I can ask a question around that topic over there.

**Comment by konstruktoid** [comment]
Interesting about `noexec` in 18.04 and you're correct, however I added the Pre/Post dpkg configuration "just in case": https://github.com/konstruktoid/hardening/commit/f86b0c1233a1850e02f03f96a11bcc10b8ce8a90

But I haven't seen anything regarding when this behavior changed.

**Comment by frederikbosch** [comment]
@konstruktoid Did you test commit f86b0c1? When I was investigating I tried to execute the `mount` commands manually on my Ubuntu 18.04 system. That failed because `/tmp` and `/var/tmp` are not in fstab. I did not find out how to (re)mount a tmpfs mounted through systemd.

**Comment by frederikbosch** [comment]
Fixed by #30
